File: leetcode/dp/2189_Number_of_Ways_to_Build_House_of_Cards/codes/dp.c
#include <string.h>

static int memo[501][168];

int dp(int rem, int cap) {
    if (rem == 0) return 1;
    if (cap <= 0) return 0;
    if (memo[rem][cap] != -1) return memo[rem][cap];
    int limit = cap;
    int t = (rem + 1) / 3;
    if (limit > t) limit = t;
    if (limit == 0) return memo[rem][cap] = 0;
    int ans = 0;
    for (int k = 1; k <= limit; k++) {
        ans += dp(rem - (3 * k - 1), k - 1);
    }
    return memo[rem][cap] = ans;
}

int houseOfCards(int n) {
    int maxK = (n + 1) / 3;
    for (int i = 0; i <= n; i++)
        for (int j = 0; j <= maxK; j++)
            memo[i][j] = -1;
    return dp(n, maxK);
}

File: leetcode/dp/2189_Number_of_Ways_to_Build_House_of_Cards/codes/dp.py
from functools import lru_cache

class Solution:
    def houseOfCards(self, n: int) -> int:
        maxK = (n + 1) // 3

        @lru_cache(None)
        def dp(rem: int, cap: int) -> int:
            if rem == 0:
                return 1
            if cap <= 0:
                return 0
            limit = min(cap, (rem + 1) // 3)
            if limit == 0:
                return 0
            ans = 0
            for k in range(1, limit + 1):
                ans += dp(rem - (3 * k - 1), k - 1)
            return ans
        
        return dp(n, maxK)
